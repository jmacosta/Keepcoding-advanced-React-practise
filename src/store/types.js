export const AUTH_LOGIN_REQUEST = 'auth/login/request';
export const AUTH_LOGIN_SUCCESS = 'auth/login/success';
export const AUTH_LOGIN_FAILURE = 'auth/login/failure';

export const UI_RESET_ERROR = 'ui/reset';

export const AUTH_LOGOUT = 'auth/logout';

export const ADVERTS_LOADED_REQUEST = 'adverts/loaded/request';
export const ADVERTS_LOADED_SUCCESS = 'adverts/loaded/success';
export const ADVERTS_LOADED_FAILURE = 'adverts/loaded/failure';

export const ADVERTS_DETAIL_REQUEST = 'adverts/detail/request';
export const ADVERTS_DETAIL_SUCCESS = 'adverts/detail/success';
export const ADVERTS_DETAIL_FAILURE = 'adverts/detail/failure';

export const ADVERTS_CREATED_REQUEST = 'adverts/created/request';
export const ADVERTS_CREATED_SUCCESS = 'adverts/created/success';
export const ADVERTS_CREATED_FAILURE = 'adverts/created/failure';

export const ADVERTS_CREATED = 'adverts/created';
