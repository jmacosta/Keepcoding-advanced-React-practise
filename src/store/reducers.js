import {
  ADVERTS_CREATED,
  ADVERTS_LOADED,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,
  UI_RESET_ERROR
} from './types';
const defaultState = {
  auth: false,
  adverts: [],
  ui: { isFetching: false, error: null }
};

export const auth = (state = defaultState.auth, action) => {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      return true;
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
};

export const adverts = (state = defaultState.adverts, action) => {
  switch (action.type) {
    case ADVERTS_LOADED:
      return action.payload;
    case ADVERTS_CREATED:

    default:
      return state;
  }
};

export const ui = (state = defaultState.ui, action) => {
  if (action.error) {
    return { isFetching: false, error: action.payload };
  }
  if (action.type.endsWith('/request')) {
    return { isFetching: true, error: null };
  }
  if (action.type.endsWith('/success')) {
    return { isFetching: false, error: null };
  }
  if (action.type === UI_RESET_ERROR) {
    return { ...state, error: null };
  }
  return state;
};
