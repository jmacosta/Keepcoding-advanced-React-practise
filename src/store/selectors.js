export const getIsLogged = state => state.auth;
export const getAdverts = state => state.adverts.data;
export const getAdvertById = advertId => state =>
  getAdverts(state).find(advert => advert.id === advertId);
export const getUi = state => state.ui;
export const areAdvertsLoaded = state => state.adverts.areLoaded;
export const areTagsLoaded = state => state.tags.areLoaded;
