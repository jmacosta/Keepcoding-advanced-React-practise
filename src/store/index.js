import { combineReducers, createStore } from 'redux';
import * as reducers from './reducers.js';
export default function configureStore(preloadedState) {
  const store = createStore(
    combineReducers(reducers),
    preloadedState,
    window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
  );
  return store;
}
