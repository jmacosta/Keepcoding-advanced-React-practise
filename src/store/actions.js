import { getLatestAdverts } from '../api/service';
import { login } from '../pages/auth/service';
import { areAdvertsLoaded } from './selectors';
import {
  ADVERTS_LOADED_FAILURE,
  ADVERTS_LOADED_REQUEST,
  ADVERTS_LOADED_SUCCESS,
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,
  UI_RESET_ERROR
} from './types';

export const authLoginSuccess = () => ({
  type: AUTH_LOGIN_SUCCESS
});

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST
});

export const authLoginFailure = error => ({
  type: AUTH_LOGIN_FAILURE,
  error: true,
  payload: error
});

export const advertsLoadedSuccess = adverts => ({
  type: ADVERTS_LOADED_SUCCESS,
  payload: adverts
});

export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST
});

export const advertsLoadedFailure = error => ({
  type: ADVERTS_LOADED_FAILURE,
  error: true,
  payload: error
});

export const authLogin = credentials => {
  return async function (dispatch, getstate) {
    try {
      dispatch(authLoginRequest());
      await login(credentials);
      dispatch(authLoginSuccess());
    } catch (error) {
      dispatch(authLoginFailure(error));
      throw error;
    }
  };
};

export const uiResetError = () => ({
  type: UI_RESET_ERROR
});

export const authLogout = () => ({
  type: AUTH_LOGOUT
});

export const loadAdverts = () => {
  return async function (dispatch, getstate) {
    if (areAdvertsLoaded(getstate())) {
      return;
    }
    try {
      dispatch(advertsLoadedRequest());
      const adverts = await getLatestAdverts();
      dispatch(advertsLoadedSuccess(adverts));
    } catch (error) {
      dispatch(advertsLoadedFailure(error));
      throw error;
    }
  };
};
