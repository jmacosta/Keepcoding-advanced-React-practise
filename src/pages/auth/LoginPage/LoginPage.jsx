import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useLocation, useNavigate } from 'react-router-dom';
import { Button } from '../../../Components/sharedComponents/Button';

import Logo from '../../../assets/logo_portrait.svg?react';
import {
  authLoginFailure,
  authLoginRequest,
  authLoginSuccess,
  uiResetError
} from '../../../store/actions';
import { getUi } from '../../../store/selectors';
import { login } from '../service';
import AtIcon from './components/AtIcon';
import LockIcon from './components/LockIcon';
import './login.css';

function LoginPage() {
  const iconOptions = {
    fill: '#2e2e2e',
    height: '16',
    width: '16',
    class: 'inputIcon'
  };
  const [credentials, setCredentials] = useState({
    email: '',
    password: '',
    rememberMe: ''
  });

  const dispatch = useDispatch();
  const { error, isFetching } = useSelector(getUi);
  const disabled = !(credentials.email && credentials.password && !isFetching);
  const location = useLocation();
  const navigate = useNavigate();
  const to = location?.state?.from || '/';
  const handleSubmit = async event => {
    event.preventDefault();
    try {
      dispatch(authLoginRequest());
      await login(credentials);
      dispatch(authLoginSuccess());
      navigate(to, { replace: true });
    } catch (error) {
      dispatch(authLoginFailure(error));
    }
  };
  const handleCredentialsChange = event => {
    setCredentials({
      ...credentials,
      [event.target.name]: event.target.value
    });
  };
  const handleRememberMeChange = event => {
    setCredentials({
      ...credentials,
      [event.target.name]: event.target.checked
    });
  };
  const resetError = () => {
    dispatch(uiResetError());
  };

  return (
    <main className='MainLoginPage'>
      <form
        className='form_main'
        action=''
        id='form-login'
        onSubmit={handleSubmit}
      >
        <div className='form_main__logo'>
          <Logo height={'10vh'} />
        </div>
        <p className='heading'>Iniciar sesiÃ³n</p>

        <div className='inputContainer'>
          <AtIcon iconOptions={iconOptions} />
          <input
            placeholder='Email'
            id='email'
            className='inputField'
            type='email'
            name='email'
            onChange={handleCredentialsChange}
            value={credentials.email}
          />
        </div>

        <div className='inputContainer'>
          <LockIcon iconOptions={iconOptions} />
          <input
            placeholder='Password'
            id='password'
            className='inputField'
            type='password'
            name='password'
            onChange={handleCredentialsChange}
            value={credentials.password}
          />
        </div>
        <Button type='submit' className='submitButton' disabled={disabled}>
          {isFetching ? 'Connecting...' : 'Enviar'}
        </Button>

        <div className='signupContainer'>
          <label>
            <input
              type='checkbox'
              name='rememberMe'
              onChange={handleRememberMeChange}
              checked={credentials.rememberMe}
            />
            RecuÃ©rdame en este dispositivo
          </label>
        </div>
        {error && (
          <div className='error' onClick={resetError}>
            ðŸš« {error.message}
          </div>
        )}
      </form>
    </main>
  );
}
export default LoginPage;
