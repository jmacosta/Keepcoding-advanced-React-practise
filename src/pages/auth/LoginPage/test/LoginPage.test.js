import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { authLogin } from '../../../../store/actions';
import LoginPage from '../LoginPage';
jest.mock('../../../../store/actions');
describe('LoginPage', () => {
  const state = { ui: { isFetching: false, error: null } };
  const store = {
    getState: () => state,
    subscribe: () => {}
  };
  const renderComponent = () =>
    render(
      <Provider store={store}>
        <BrowserRouter>
          <LoginPage />
        </BrowserRouter>
      </Provider>
    );

  test('snapshot', () => {
    const result = renderComponent();

    expect(result.container).toMatchSnapshot();
  });
  test('Should dispatch authLogin action', () => {
    const email = 'user@example.com';
    const password = '123456';
    const rememberMe = '';

    renderComponent();
    const emailInput = screen.getByPlaceholderText(/Email/);
    const passwordInput = screen.getByPlaceholderText(/Password/);

    const submitButton = screen.getByRole('button');
    expect(submitButton).toBeDisabled;
    fireEvent.change(emailInput, { target: { value: email } });
    fireEvent.change(passwordInput, { target: { value: password } });

    fireEvent.click(submitButton);
    expect(submitButton).toBeEnabled;
    expect(authLogin).toHaveBeenCalledWith({ email, password, rememberMe });
  });
});
