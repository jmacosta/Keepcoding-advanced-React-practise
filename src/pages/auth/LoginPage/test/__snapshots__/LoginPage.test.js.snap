// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <main
    class="MainLoginPage"
  >
    <form
      action=""
      class="form_main"
      id="form-login"
    >
      <div
        class="form_main__logo"
      />
      <p
        class="heading"
      >
        Iniciar sesión
      </p>
      <div
        class="inputContainer"
      >
        <input
          class="inputField"
          id="email"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
      <div
        class="inputContainer"
      >
        <input
          class="inputField"
          id="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <button
        class="submitButton"
        disabled=""
        type="submit"
      >
        Enviar
      </button>
      <div
        class="signupContainer"
      >
        <label>
          <input
            name="rememberMe"
            type="checkbox"
          />
          Recuérdame en este dispositivo
        </label>
      </div>
    </form>
  </main>
</div>
`;
